src/AMAMM.sol

```diff
@@ -2,69 +2,176 @@
-                beforeDonate: false,
-                afterDonate: false,
-                beforeSwapReturnDelta: false,
-                afterSwapReturnDelta: true, // Allow afterSwap to return a custom delta
-                afterAddLiquidityReturnDelta: false,
-                afterRemoveLiquidityReturnDelta: false
-            });
+    /// -----------------------------------------------------------------------
+    /// Library usage
+    /// -----------------------------------------------------------------------
+
+    using SafeCastLib for *;
+    using FixedPointMathLib for *;
+
+    /// -----------------------------------------------------------------------
+    /// Constants
+    /// -----------------------------------------------------------------------
+
+    function K(PoolId) internal view virtual returns (uint40) {
+        return 24;
+    }
+
+    function EPOCH_SIZE(PoolId) internal view virtual returns (uint256) {
+        return 1 hours;
+    }
+
+    function MIN_BID_MULTIPLIER(PoolId) internal view virtual returns (uint256) {
+        return 1.1e18;
+    }
+
+    /// -----------------------------------------------------------------------
+    /// Storage variables
+    /// -----------------------------------------------------------------------
+
+    mapping(PoolId id => uint40) internal _lastUpdatedEpoch;
+    mapping(address deposits => uint256) public _userBalance;
+    mapping(Currency currency => uint256) internal _totalFees;
+    mapping(PoolId id => mapping(uint40 => Bid)) public poolEpochManager;
+    mapping(address manager => mapping(Currency currency => uint256)) internal _fees;
+
+    /// -----------------------------------------------------------------------
+    /// Getter actions
+    /// -----------------------------------------------------------------------
+
+    function getManager(PoolId id, uint40 epoch) public view returns (Bid memory) {
+        return poolEpochManager[id][epoch];
+    }
+
+    /// -----------------------------------------------------------------------
+    /// Bidder actions
+    /// -----------------------------------------------------------------------
+    /// @inheritdoc IAmAmm
+    function bid(PoolId id, bytes7 payload, uint128 rent, uint128 deposit, uint40 _epoch) external virtual override {
+        /// -----------------------------------------------------------------------
+        /// Validation
+        /// -----------------------------------------------------------------------
+
+        address msgSender = LibMulticaller.senderOrSigner();
+
+        // ensure bid is valid
+        // - manager can't be zero address
+        // - bid needs to be greater than the next bid by >10%
+        // - deposit needs to cover the rent for K hours
+        // - deposit needs to be a multiple of rent
+        // - payload needs to be valid
+        if (
+            rent <= poolEpochManager[id][_epoch].rent.mulWad(MIN_BID_MULTIPLIER(id)) || deposit < rent * K(id)
```

Comment:
@leckylao Im not sure what you mean, rent calculation is done in `_updateEpochBids`.
Or are you suggesting that we increase the "deposit" from what the user selects in the param to make sure tx doesn't revert? 

This would mean that if I want to bid 1ETH my actual bid can be 2ETH because the function covered from my previous balances.